kroxylicious-additional-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/BufferPool.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.filter.encryption;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+
+interface BufferPool {
+
+    ByteBuffer acquire(int size);
+
+    void release(ByteBuffer buffer);
+
+    static BufferPool allocating() {
+        return new BufferPool() {
+            @Override
+            public ByteBuffer acquire(int size) {
+                return ByteBuffer.allocate(size);
+            }
+
+            @Override
+            public void release(ByteBuffer buffer) {
+
+            }
+        };
+    }
+
+    static BufferPool directAllocating() {
+        return new BufferPool() {
+            @Override
+            public ByteBuffer acquire(int size) {
+                return ByteBuffer.allocateDirect(size);
+            }
+
+            @Override
+            public void release(ByteBuffer buffer) {
+
+            }
+        };
+    }
+
+    static BufferPool pooled(int poolSize) {
+        AtomicReferenceArray<ArrayBlockingQueue<ByteBuffer>> freeLists = new AtomicReferenceArray<>(32);
+
+        return new BufferPool() {
+            @Override
+            public ByteBuffer acquire(int size) {
+                int index = ceil(size);
+                ArrayBlockingQueue<ByteBuffer> freeList = freeList(index);
+                return freeList.poll();
```

Comment:
presumably you mean to handle the case when freeList.poll() returns null and actually allocate a buffer.

Reply:
The whole BufferPool thing is very much a WIP at the moment.

Reply:
I took out all by the allocating pool for the moment. Since this is a performance optimization it seems like a good idea to focus on the functionality for now.